### PR TITLE
Use fork weight instead of individual bank weight for fork selection.

### PR DESCRIPTION
**Pull request by aeyakovenko**
#### Summary of Changes

Fixes #

tag: @carllin

### CODE REVIEW

File: core/src/consensus.rs [hunk 1]
```diff
@@ -67,12 +67,13 @@ impl Tower {
...
         vote_accounts: F,
         ancestors: &HashMap<Slot, HashSet<u64>>,
-    ) -> (HashMap<Slot, StakeLockout>, u64)
+    ) -> (HashMap<Slot, StakeLockout>, u64, u128)
     where
         F: Iterator<Item = (Pubkey, (u64, Account))>,
     {
         let mut stake_lockouts = HashMap::new();
         let mut total_stake = 0;
+        let mut total_weight = 0;
```
Comment (garious) [inline core/src/consensus.rs]: I think what you're calculating here is what's already called `commitment`. Any way you can reuse that code?

Comment (aeyakovenko) [reply core/src/consensus.rs]: Weight is a better word for consensus.  Heaviest chain rule.

Comment (garious) [reply core/src/consensus.rs]: I'm not asking you to use the term `commitment`. I'm asking you to use the existing commitment calculation.

fyi:
```
git grep commitment | wc -l
     560
```

Comment (aeyakovenko) [reply core/src/consensus.rs]: it doesn't look like its computing the same thing, core/src/commitment.rs

File: core/src/replay_stage.rs [hunk 2]
```diff
@@ -770,28 +806,44 @@ impl ReplayStage {
...
                     );
                     Self::confirm_forks(tower, &stake_lockouts, total_staked, progress, bank_forks);
                     stats.total_staked = total_staked;
-                    stats.weight = tower.calculate_weight(&stake_lockouts);
+                    stats.weight = bank_weight;
+                    stats.fork_weight = stats.weight
+                        + bank
+                            .parent()
+                            .and_then(|b| progress.get(&b.slot()))
+                            .map(|x| x.fork_stats.fork_weight)
```
Comment (carllin) [inline core/src/replay_stage.rs]: can we hard assert here that the fork stats must exist in the `progress` map

File: core/src/replay_stage.rs [hunk 3]
```diff
@@ -811,25 +863,11 @@ impl ReplayStage {
...
-            let mut parents: Vec<_> = if let Some(set) = ancestors.get(&stats.slot) {
-                set.iter().collect()
-            } else {
-                vec![]
-            };
-            parents.sort();
-            debug!("{}: {:?} {:?}", stats.slot, stats, parents,);
-        });
-        let rv = Self::pick_best_fork(ancestors, &candidates);
+        candidates.sort_by_key(|b| (b.1.fork_weight, 0i64 - b.1.slot as i64));
```
Comment (carllin) [inline core/src/replay_stage.rs]: Can we add a few tests here to illustrate the expected behavior in scenarios where parent-child fork weights are the same? If we are sure that parent weight > child weight in all scenarios, can we assert above that `bank_weight + fork_weight > fork_weight`, i.e. `bank_weight > 0` ( I thnk this is true b/c we always at last simulate one vote)

File: local-cluster/tests/local_cluster.rs [hunk 4]
```diff
@@ -200,7 +200,7 @@ fn run_network_partition(partitions: &[usize]) {
         ..ClusterConfig::default()
     };
     let now = timestamp();
-    let partition_start = now + 30_000;
+    let partition_start = now + 60_000;
```
Comment (carllin) [inline local-cluster/tests/local_cluster.rs]: Is the goal here to start the partition after all the validators are already in the leader schedule? If so, there's some logic in `cluster_tests::sleep_n_epochs` that can more precisely calculate how long to wait based on epoch len, ticks per slot, etc.

File: core/src/replay_stage.rs (hunk 3 above)
Comment (aeyakovenko) [reply core/src/replay_stage.rs]: https://github.com/solana-labs/solana/issues/7085

File: local-cluster/tests/local_cluster.rs (hunk 4 above)
Comment (aeyakovenko) [reply local-cluster/tests/local_cluster.rs]: https://github.com/solana-labs/solana/issues/7086

**Comment by aeyakovenko** [comment]
security issue: https://github.com/solana-labs/solana/issues/7087

**Pull request merged**